dd-sdk-android/src/main/kotlin/com/datadog/android/v2/core/internal/storage/ConsentAwareStorage.kt

```diff
@@ -0,0 +1,111 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.v2.core.internal.storage
+
+import com.datadog.android.core.internal.persistence.file.FileHandler
+import com.datadog.android.core.internal.persistence.file.FileOrchestrator
+import com.datadog.android.privacy.TrackingConsent
+import com.datadog.android.v2.api.BatchWriterListener
+import com.datadog.android.v2.api.InternalLogger
+import com.datadog.android.v2.api.context.DatadogContext
+import java.io.File
+import java.util.Locale
+
+internal class ConsentAwareStorage(
+    private val grantedOrchestrator: FileOrchestrator,
+    private val pendingOrchestrator: FileOrchestrator,
+    private val handler: FileHandler,
+    private val listener: BatchWriterListener,
+    private val internalLogger: InternalLogger
+) : Storage {
+
+    /**
+     * Keeps track of files currently being read.
+     */
+    private val lockedFiles: MutableSet<File> = mutableSetOf()
+
+    /** @inheritdoc */
+    override fun writeCurrentBatch(
+        datadogContext: DatadogContext,
+        callback: (BatchWriter) -> Unit
+    ) {
+        val orchestrator = when (datadogContext.trackingConsent) {
+            TrackingConsent.GRANTED -> grantedOrchestrator
+            TrackingConsent.PENDING -> pendingOrchestrator
+            TrackingConsent.NOT_GRANTED -> null
+        }
+
+        val writer = object : BatchWriter {
+            override fun currentMetadata(): ByteArray? {
+                // TODO RUMM-2186 handle writing/updating batch metadata in separate file
+                return null
+            }
+
+            override fun write(event: ByteArray, eventId: String, newMetadata: ByteArray) {
+                // prevent useless operation for empty event / null orchestrator
+                if (event.isEmpty() || orchestrator == null) {
+                    listener.onDataWritten(eventId)
+                    return
+                }
+
+                val file = orchestrator.getWritableFile(event.size)
```

Comment:
> (...) by the time it is invoked we may have other consent -> different orchestrator, not the one we captured for invocation

Not sure if I understand this @0xnm 🤔💭. We get the orchestrator based on the immutable `datadogContext` passed as an parameter to `writeCurrentBatch()`. Then, any time later we receive a callback which provides us API (`BatchWriter`) to writing data. The write strategy will respect the consent value from `datadogContext`.

From the caller's POV, IMO everything should be consistent with this implementation: the caller orders write for given consent value and it knows that `BatchWriter` will respect it. No?

Reply:
I mean that generally the API contract is that this callback may be invoked in async manner (if I got it right). My point is that what if user changed a consent in the timeframe between the moment we received an `orchestrator` and the moment we use it in the `BatchWriter`. Then event may be written to the location which belongs to the old consent.

This is quite hypothetical case (and may be quite an edge case), but this gist of it is that imo we need to keep in mind the possible SDK state changes with async invocations.

Reply:
This is an edge case that is not something we want to dig into. The time frame for such a concurrency issue will on most case something smaller than 500ms, which is totally acceptable.

Reply:
I get the point now @0xnm 👌. Agree, it's kind of edge case, but it can be clearly visible in such user situation:
```swift
// all synchronous, on a single thread:
Datadog.set(trackingConsent: .granted)
logger.info("granted log")
Datadog.set(trackingConsent: .pending)
```
In such case, the _"granted log"_ should be sent by the SDK. This is however a question on how we will further synchronise things in higher implementation (even higher than `Storage` level). The RFC recommends processing all user calls through context thread first (see [_Threading in `DDCore`_](https://datadoghq.atlassian.net/wiki/spaces/RUMP/pages/2344748708/RFC+-+V2+Core+Storage+and+Batching#Threading-in-DDCore)) to support exactly these edge cases.